### DISCUSSION TITLE
How to cancel timed out queueBind

### DISCUSSION THREAD

**Discussion by MartinKosicky** [rabbitmq/rabbitmq-dotnet-client#1079] [General]
Hi my question is how do I recover from situation where the 
   
`_channel.QueueBind(`
returns timeout for some reason. Because for my situation, if i do many queueBind in a row and if one times out, the next returns Pipelining is not supported. How do I recover from such situation?

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Comment by michaelklishin** [comment]
Open a new channel as "pipelining is not supported" is likely a [channel-level exception](https://www.rabbitmq.com/channels.html#error-handling).

**Comment by MartinKosicky** [comment]
I have a code here that shows the problem:  

First Bind takes long to finish, then there is a timeout and then a bunch of pipeline failures and then some success (I guess it somehow got recovered). I would like to understand this behaviour. (For me it seems like some side thread doing all the work, and there is some non blocking swap which must be first consumed until I can do another call)

RabbitMq is started on my virtualbox machine with docker 

`
docker run -d  --name rabbit-server -e RABBITMQ_DEFAULT_USER=admin -e RABBITMQ_DEFAULT_PASS=admin -p 5672:5672 -p 15672:15672
`

```
        static async Task Main(string[] args)
        {
            var factory = new ConnectionFactory()
            {
                HostName = "127.0.0.1",
                UserName = "admin",
                Password = "admin",
            };
            factory.DispatchConsumersAsync = true;
            var _connection = factory.CreateConnection();
            var _channel = _connection.CreateModel();

            _channel.ExchangeDeclare(exchange: "ex1", type: ExchangeType.Direct, autoDelete: false, durable: true);

            _channel.QueueDeclare(queue: "req1", durable: true, exclusive: true, autoDelete: false, arguments: null);
            string _replyQueue = _channel.QueueDeclare().QueueName;
            object o = new object();

            List<Task> alltasks = new List<Task>();
            for (int i=0;i<500;i++)
            {
                int z = i;
                Console.WriteLine($"Binding {i}");
                var t = Task.Run(() =>
                {
                    lock(o)
                    {
                        try
                        {
                            Console.WriteLine($"Starting bind {z}");
                            _channel.QueueBind("req1", "ex1", "rk" + z);
                        } catch (Exception ex)
                        {
                            Console.WriteLine($"Failed to bind {z} because {ex}");
                        }
                        
                    }                    
                    Console.WriteLine($"Bind {z} finished");
                });
                alltasks.Add(t);


            }

            foreach (Task t in alltasks)
            {
                await t;
            }

            Console.WriteLine("Hello World!");
        }
```

**Comment by MartinKosicky** [comment]
Ok I realised that I must use SemaphoreSlim , probabbly the RabbitMQ async worker is using the same thread pool. Perhaps it might be good to mention in the documentation regarding sharing channel between threads to use  SemaphoreSlim and not lock in case of async programs
